### PR TITLE
Document the `@noescape` SIL function type attribute.

### PR DESCRIPTION
**Pull request by atrick**

### CODE REVIEW

File: docs/SIL.rst [hunk 1]
```diff
@@ -437,7 +437,13 @@ number of ways:
     a guaranteed direct parameter.
 
   - Otherwise, the context value is treated as an unowned direct
-    parameter.
+    parameter. FIXME!!! what is this?? It sounds like ``@noescape
```
Comment (aschwaighofer) [inline docs/SIL.rst]: This predate ``@noescape``. ``@convention(block)``'s context is passed unowned

From test/IRGen/objc_block.sil:

```
// CHECK-LABEL: define{{( protected)?}} swiftcc %T10objc_block3FooC* @call_block(%objc_block*, %T10objc_block3FooC*) {{.*}} {
// CHECK:       entry:
// CHECK:         [[T0:%.*]] = getelementptr inbounds %objc_block, %objc_block* %0, i32 0, i32 3
// CHECK:         [[T1:%.*]] = load i8*, i8** [[T0]]
// CHECK:         [[T2:%.*]] = bitcast i8* [[T1]] to i8* (%objc_block*, i8*)*
// CHECK:         [[T3:%.*]] = bitcast %T10objc_block3FooC* %1 to i8*
// CHECK:         [[T4:%.*]] = call i8* [[T2]](%objc_block* %0, i8* [[T3]])
// CHECK:         [[T5:%.*]] = bitcast i8* [[T4]] to %T10objc_block3FooC*
// CHECK:         ret %T10objc_block3FooC* [[T5]]
// CHECK:       }
```

**Comment by atrick** [comment]
@aschwaighofer is it fair to say that ``@callee_x`` attribute implies ``@convention(thick)`` attribute?

**Comment by aschwaighofer** [comment]
It only makes sense to have an ownership qualifier for the context on a function representation that has a context. Yes I think so.

**Comment by aschwaighofer** [comment]
We can have:

- @convention(thin) () -> ()
No context, can be converted to a @convention(thin) @noescape type.

- @convention(block) () -> ()
Objective-C block convention; context is passed @unowned in spirit.

- Thick swift function types: @callee_guaranteed () -> ()
Can be @noescape in which case the context value is treated as a trivial type -- which admittedly is a little odd because we still spell out ownership (which we usually only do for reference counted things).
@callee_owned is dead.

**Comment by atrick** [comment]
@aschwaighofer here's another revision. I'll let you figure out how to kill off `@callee_owned`. It's still all over the tests and handled in ParseDecl.cpp.

**Comment by aschwaighofer** [comment]
This looks good to me.

**Comment by atrick** [comment]
@swift-ci smoke test and merge.

**Comment by atrick** [comment]
@swift-ci smoke test and merge.

**Pull request merged**